### ISSUE TITLE
Lost connection doesn't terminate process

### ISSUE THREAD

**Issue by pbrisbin** [frontrowed/faktory_worker_haskell#17]
Twice now in production, we've experienced that restarting the Faktory server doesn't cause consumer processes to crash. We see they have trouble with the socket:

```
...: Network.Socket.sendBuf: resource vanished (Broken pipe)
```

But the process stays up -- though it doesn't accept any Jobs. It also has very high CPU when in this state. (We also observed issues producing Jobs to Faktory while such consumers were running, but I wasn't able to reproduce this in isolation.)

This is bad because nothing replaces the now-broken service and no work is being done until a deployment happens to "kick" the service. This can be very bad if that work was meant to happen in a quiet time and the deployment causes it to pick up again in a busy time.

## Reproduction Steps

1. Run faktory (via docker)
1. Run consumer example
1. Restart faktory
1. Observe:

   - The process sees errors but doesn't crash

     ```
     Starting consumer loop
     [ERROR]: SHUTDOWN Shutdown in progress
     [ERROR]: No more input
     faktory-example-consumer: Network.Socket.sendBuf: resource vanished (Broken pipe)
     ```

   - Jobs (produced with the `example-producer`) are ignored

   - Also, my laptop started going crazy with this processing take 100% CPU.

## Expected Outcome

These errors should crash the consumer process.

/cc @eborden @jdreaver

**Comment by eborden** [comment]
PR incoming...

**Comment by jdreaver** [comment]
Oh wow I didn't see any high CPU after the socket breaking in our prod systems. Good to know that's a possible cause of high CPU though, because I have seen it other times without a good explanation.

**Comment by eborden** [comment]
Resolved by #18
